MatrixSDK/Crypto/Dehydration/MXDehydrationManager.m

```diff
@@ -0,0 +1,178 @@
+        return;
+    }
+    
+    MXKeyData * keyData =  [[MXKeyProvider sharedInstance] keyDataForDataOfType:MXSessionDehydrationKeyDataType isMandatory:NO expectedKeyType:kRawData];
+    if (!keyData)
+    {
+        MXLogDebug(@"[MXDehydrationManager] dehydrateDevice: No dehydrated key.");
+        success(nil);
+        return;
+    }
+    NSData *key = ((MXRawDataKey*) keyData).key;
+
+    inProgress = YES;
+    
+    OLMAccount *account = [[OLMAccount alloc] initNewAccount];
+    NSDictionary *e2eKeys = [account identityKeys];
+
+    NSUInteger maxKeys = [account maxOneTimeKeys];
+    [account generateOneTimeKeys:maxKeys / 2];
+
+    // [account account.generateFallbackKey];
+    [account markOneTimeKeysAsPublished];
+    
+    MXLogDebug(@"[MXDehydrationManager] dehydrateDevice: account created %@", account.identityKeys);
+    
+    // dehydrate the account and store it on the server
+    NSError *error = nil;
+    MXDehydratedDevice *dehydratedDevice = [MXDehydratedDevice new];
+    dehydratedDevice.account = [account serializeDataWithKey:key error:&error];
+    dehydratedDevice.algorithm = MXDehydrationAlgorithm;
+    
+    if (error)
+    {
+        inProgress = NO;
+        MXLogError(@"[MXDehydrationManager] dehydrateDevice: account serialization failed: %@", error);
+        failure(error);
+        return;
+    }
+    
+    [crypto.matrixRestClient setDehydratedDevice:dehydratedDevice withDisplayName:@"Backup device" success:^(NSString *deviceId) {
+        MXLogDebug(@"[MXDehydrationManager] dehydrateDevice: preparing device keys for device %@ (current device ID %@)", deviceId, crypto.myDevice.deviceId);
+        MXDeviceInfo *deviceInfo = [[MXDeviceInfo alloc] initWithDeviceId:deviceId];
+        deviceInfo.userId = self->crypto.matrixRestClient.credentials.userId;
+        deviceInfo.keys = @{
+            [NSString stringWithFormat:@"ed25519:%@", deviceId]: e2eKeys[@"ed25519"],
+            [NSString stringWithFormat:@"curve25519:%@", deviceId]: e2eKeys[@"curve25519"]
+        };
+        deviceInfo.algorithms = [[MXCryptoAlgorithms sharedAlgorithms] supportedAlgorithms];
+        
+//        NSDictionary *deviceKeys = @{
+//            @"algorithms": [[MXCryptoAlgorithms sharedAlgorithms] supportedAlgorithms],
+//            @"device_id": deviceId,
+//            @"user_id": crypto.mxSession.myUserId,
+//            @"keys": @{
+//                    [NSString stringWithFormat:@"ed25519:%@", deviceId]: account.identityKeys[@"ed25519"],
+//                    [NSString stringWithFormat:@"curve25519:%@", deviceId]: account.identityKeys[@"curve25519"]
+//            }
+//        };
+//
+//        NSString *deviceSignature = [account signMessage:[NSJSONSerialization dataWithJSONObject:deviceKeys options:0 error:&error]];
+
+
+        NSError *error = nil;
+        NSString *signature = [account signMessage:[NSJSONSerialization dataWithJSONObject:deviceInfo.signalableJSONDictionary options:0 error:&error]];
```

Comment:
Signatures need to be done using canonical JSON, like in  https://github.com/matrix-org/matrix-ios-sdk/blob/develop/MatrixSDK/Crypto/MXOlmDevice.m#L116 (or you can use the `signJSON` method)